src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -877,7 +901,7 @@ private TypeGenerationSpec GetOrAddTypeGenerationSpec(Type type, JsonSourceGener
                     if (!type.TryGetDeserializationConstructor(useDefaultCtorInAnnotatedStructs, out ConstructorInfo? constructor))
                     {
                         classType = ClassType.TypeUnsupportedBySourceGen;
-                        _sourceGenerationContext.ReportDiagnostic(Diagnostic.Create(MultipleJsonConstructorAttribute, Location.None, new string[] { $"{type}" }));
+                        _typeLevelDiagnostics.Add((type, MultipleJsonConstructorAttribute, new string[] { $"{type}" }));
```

Comment:
Out of curiosity, what necessitated the change from reporting the diagnostic directly to storing in a set? Were there duplicates? 

Probably not in scope for this PR, but in future iterations I would recommend doing the same for emitted member-level diagnostics (in other words, we should try to make the mapping layer pure).

Reply:
Changed to a list, since uniqueness is not a consideration here. The reason for the change is that we sometimes want the location of the diagnostic to be the `[JsonSerializable]` attribute application. Because a type might be encountered either from the attribute or from a nested object graph, and because we can't determine the order in which we encounter the type, we wait till all types and attribute applications are processed so that we can get the attribute application if specified before emitting the diagnostic.